rust/lance/src/dataset/builder.rs

```diff
@@ -110,6 +112,13 @@ impl DatasetBuilder {
         self
     }
 
+    /// Sets `version` for the builder using a branch and optional version number
+    /// If version_number is null, checkout the latest version
+    pub fn with_branch(mut self, branch: &str, version_number: Option<u64>) -> Self {
```

Comment:
We might deside if we really need this option as well as with_tag.

This reference option might cause double loading if the uri is not the branch we want to checkout. But this is not a technical trouble but a design concern. I implemented the double loading below